File: webmvp/src/lib/chordMarks.ts
import type { ChordMark, LyricLine, Section } from "@/lib/types";

export function getMarkStart(mark: ChordMark): number {
  if (typeof mark.start === "number") {
    return mark.start;
  }
  return mark.position ?? 0;
}

export function getMarkEnd(mark: ChordMark): number {
  if (typeof mark.end === "number") {
    return mark.end;
  }
  const start = getMarkStart(mark);
  return start + 1;
}

export function normalizeChordMark(mark: ChordMark): ChordMark {
  const start = getMarkStart(mark);
  const end = Math.max(getMarkEnd(mark), start + 1);
  return {
    chord: mark.chord,
    start,
    end,
  };
}

/** Keep start/end within `[0, lyrics.length]` (fixes trailing ChordPro marks and bad data). */
export function clampChordMarkToLyrics(
  mark: ChordMark,
  lyricLength: number,
): ChordMark {
  const normalized = normalizeChordMark(mark);

  if (lyricLength <= 0) {
    return { chord: normalized.chord, start: 0, end: 1 };
  }

  let { start, end } = normalized;

  if (end > lyricLength) {
    if (start >= lyricLength) {
      start = lyricLength - 1;
      end = lyricLength;
    } else {
      end = lyricLength;
    }
  }

  if (start > lyricLength) {
    start = Math.max(0, lyricLength - 1);
  }

  if (start >= end) {
    end = Math.min(start + 1, lyricLength);
  }

  return normalizeChordMark({ chord: normalized.chord, start, end });
}

export function normalizeLyricLine(line: LyricLine): LyricLine {
  const lyricLength = line.lyrics.length;
  return {
    lyrics: line.lyrics,
    chords: line.chords.map((mark) =>
      clampChordMarkToLyrics(normalizeChordMark(mark), lyricLength),
    ),
  };
}

export function normalizeSections(sections: Section[]): Section[] {
  return sections.map((section) => ({
    ...section,
    lines: section.lines.map(normalizeLyricLine),
  }));
}

export function serializeChordMarkForPublish(mark: ChordMark): ChordMark {
  const normalized = normalizeChordMark(mark);
  return {
    chord: normalized.chord,
    start: normalized.start,
    end: normalized.end,
  };
}

export function serializeSectionsForPublish(sections: Section[]): Section[] {
  return sections.map((section) => ({
    ...section,
    lines: section.lines.map((line) => ({
      lyrics: line.lyrics,
      chords: line.chords.map(serializeChordMarkForPublish),
    })),
  }));
}

export function createChordMark(
  chord: string,
  start: number,
  end: number,
): ChordMark {
  return normalizeChordMark({
    chord,
    start,
    end: Math.max(end, start + 1),
  });
}

export function markKey(mark: ChordMark): string {
  const normalized = normalizeChordMark(mark);
  return `${normalized.start}:${normalized.end}:${normalized.chord}`;
}

File: mobile/lib/domain/chord_marks.dart
import 'package:lf_chords/data/models/chord_mark.dart';
import 'package:lf_chords/data/models/lyric_line.dart';
import 'package:lf_chords/data/models/section.dart';

int getMarkStart(ChordMark mark) {
  if (mark.start != null) {
    return mark.start!;
  }
  return mark.position ?? 0;
}

int getMarkEnd(ChordMark mark) {
  if (mark.end != null) {
    return mark.end!;
  }
  final start = getMarkStart(mark);
  return start + 1;
}

ChordMark normalizeChordMark(ChordMark mark) {
  final start = getMarkStart(mark);
  final end = start + 1 > getMarkEnd(mark) ? start + 1 : getMarkEnd(mark);
  return ChordMark(
    chord: mark.chord,
    start: start,
    end: end,
  );
}

ChordMark clampChordMarkToLyrics(ChordMark mark, int lyricLength) {
  final normalized = normalizeChordMark(mark);

  if (lyricLength <= 0) {
    return ChordMark(chord: normalized.chord, start: 0, end: 1);
  }

  var start = normalized.start!;
  var end = normalized.end!;

  if (end > lyricLength) {
    if (start >= lyricLength) {
      start = lyricLength - 1;
      end = lyricLength;
    } else {
      end = lyricLength;
    }
  }

  if (start > lyricLength) {
    start = lyricLength > 0 ? lyricLength - 1 : 0;
  }

  if (start >= end) {
    end = (start + 1).clamp(0, lyricLength);
  }

  return normalizeChordMark(
    ChordMark(chord: normalized.chord, start: start, end: end),
  );
}

LyricLine normalizeLyricLine(LyricLine line) {
  final lyricLength = line.lyrics.length;
  return LyricLine(
    lyrics: line.lyrics,
    chords: line.chords
        .map((m) => clampChordMarkToLyrics(normalizeChordMark(m), lyricLength))
        .toList(),
  );
}

List<Section> normalizeSections(List<Section> sections) {
  return sections
      .map(
        (section) => Section(
          label: section.label,
          lines: section.lines.map(normalizeLyricLine).toList(),
        ),
      )
      .toList();
}

String markKey(ChordMark mark) {
  final normalized = normalizeChordMark(mark);
  return '${normalized.start}:${normalized.end}:${normalized.chord}';
}
